### PR TITLE
Unique Pointers (Thread / Hart Rework)

### PR DESCRIPTION
**Pull request by rkabrick**
I added `@DAVE` (@donofrio) and `@Lee` (@leekillough) comments where I knew I needed you two to look specifically. However, feel free to tear the rest of it apart as you see fit.

## Minor Changes
- `RevCommon`:
  + Moved definition of `ThreadState` here
  + Changed `MAX_HARTS` to 4096 (This can be changed back but until we have someone that needs more than that there's no sense in carrying around those extra bits in the bitsets)

- `RevProc`:
  + Changed `HartToDecode` => `HartToDecodeID` to distinguish it is not an instance of `RevHart`
  + Changed `HartToExec` => `HartToExecID` to distinguish it is not an instance of `RevHart`
  + Made a `ActiveThreadID` variable so we didn't have to call into a function (ie. `GetActiveThreadID()`) which gets set at the same time `RegFile` does
  + Took ALUFault Injection Logic and put it in its own function (solely to clean up `RevProc::ClockTick`)
  + Changed `HARTS_CTE` and `HARTS_CTS` to `HartsClearToExecute` and `HartsClearToDecode`, respectively
  + Changed `GetHartID()` to `GetNextHartToDecodeID()` because `GetHartID()` sounds like you're getting the ID of the currently executing Hart
  + Added `HartHasNoDependencies(unsigned HartID)` which just returns the opposite of `AnyDependency`

- `RevFeature`:
  + Changed `HartToDecode` => `HartToDecodeID` to distinguish it is not an instance of `RevHart`
  + Changed `HartToExec` => `HartToExecID` to distinguish it is not an instance of `RevHart`

- `RevThread`:
  + Changed `GetThreadID` to just `GetID` because `Thread->GetThreadID()` seemed silly

- `RevSysCalls`:
  + Changed all the calls to `GetActiveThreadID()` to use the new `ActiveThreadID` variable

- `pthread_arg_passing test`:
  + This test was never updated to use the updated interface for `rev_pthread_create`

## Major Changes
**TL;DR**: `RevThread`'s are now exclusively used as `std::unique_ptr` and load their register state into a `RevHart`'s `RegFile` when assigned to that Hart. Below are the implications of this change

- `RevCPU`:
  + Removed global `Threads` table as there was no benefit to having each thread object live here
  + Consequently removed `AssignedThreads` because it really just added a layer of indirection and once again we gained nothing from having that global state of every thread that is on a specific RevProc. We can derive this information if need be and `RevCPU` won't ever be the one initiating a state change directly for a `RevThread`
  + When checking a Proc's `ThreadsThatChangedState` queue, it now transfers the entire queue to `RevCPU` as the `std::unique_ptr<RevThread>` needs to be moved anyways (No Thread will stay in that queue across multiple clock ticks)
  + Changed `ThreadQueue` to `ReadyThreads` which now holds the actual `RevThread` instead of just the ID. Conveyed the meaning more in my opinion
  + Changed `CheckForThreadStateChanges` => `HandleThreadStateChanges` because "checking" doesn't actually encompass what's happening

- `RevProc` / `RevHart`:
  + Removed `AssignedThreads` structure. This was effectively an unnecessary layer of indirection (ie. RevHart had the TID so every time you wanted to do anything on the thread you'd have to look it up in `AssignedThreads`)
  + `RevHart`'s now own the `std::unique_ptr<RevThread>` they are assigned and also have a `std::unique_ptr<RevRegFile>` which loads in the `RevThread's` virtual register state when assigned
  + New Bitsets introduced to convey more state
    - `HartsClearToDecode`: Formally `HART_CTS`
    - `HartsClearToExecute`: Formally `HART_CTE`
    - `IdleHarts`: A Hart is considered Idle if it does not have a `RevThread` assigned to it
    - `ValidHarts`: This is used to 'trim-the-fat' that is introduced by having `MAX_HARTS` by allowing for bitwise ops (ie. `if( IdleHarts == ValidHarts )` we know there are no threads assigned to this `RevProc`)
  + Removed `Utilization` as all we ever have to know is if there is an available (Idle) hart, or if there are no Idle harts
  + Added an `UpdateHartStatus` function which sets the values of `HartsClearToDecode` each `RevProc::ClockTick`
  + Removing a thread from a Hart is now done through `PopThreadFromHart` which returns the `std::unique_ptr<RevThread>` that was formally assigned to that Hart and marks the `RevHart` as Idle

- `RevThread`:
  + Changed `RegFile` member to `VirtRegState` and added a `using RevVirtRegState = RevRegFile;` to the class because software threads do not have a register file
    This is more of a semantic change, however, I believe it helps convey the other changes introduced in this PR
  + Changed the constructor to take in a `std::unique_ptr<RevVirtRegState>` instead of a whole bunch of values that get added to a regfile (StackPtr, ThreadPtr, etc.)

---

Overall this cuts the execution time of `BIG_LOOP` by ~50% and makes Rev a little closer aligned to how hardware operates.

### CODE REVIEW

File: src/RevCPU.cc [hunk 1]
```diff
@@ -770,128 +729,122 @@ void RevCPU::CheckBlockedThreads(){
...
 
 // Checks for state changes in the threads of a given processor index 'i'
 // and handle appropriately
-void RevCPU::CheckForThreadStateChanges(uint32_t ProcID){
+void RevCPU::HandleThreadStateChangesForProc(uint32_t ProcID){
   // Handle any thread state changes for this core
   // NOTE: At this point we handle EVERY thread that changed state every cycle
-  while( !Procs[ProcID]->GetThreadsThatChangedState().empty() ){
-    auto& Thread = Procs[ProcID]->GetThreadsThatChangedState().front();
+  std::queue<std::unique_ptr<RevThread>> ThreadsThatChangedState = Procs[ProcID]->TransferThreadsThatChangedState();
```
Comment (leekillough) [inline src/RevCPU.cc:771]: `std::queue<std::unique_ptr<RevThread>> ThreadsThatChangedState = ...` might be simpler as just `auto ThreadsThatChangedState = ...`

File: src/RevCPU.cc [hunk 2]
```diff
@@ -770,128 +729,122 @@ void RevCPU::CheckBlockedThreads(){
...
+  while( !ThreadsThatChangedState.empty() ){
+    std::unique_ptr<RevThread> Thread = std::move(ThreadsThatChangedState.front());
```
Comment (leekillough) [inline src/RevCPU.cc:773]: See above

File: src/RevCPU.cc [hunk 3]
```diff
@@ -770,128 +729,122 @@ void RevCPU::CheckBlockedThreads(){
...
+      Enabled[ProcID] = false;
+    }
+    // Remove the pointer to the thread as it was handled
+    ThreadsThatChangedState.pop();
   }
   return;
 }
 
+void RevCPU::InitMainThread(uint32_t MainThreadID, const uint64_t StartAddr){
+  // @Lee: Is there a better way to get the feature info?
```
Comment (leekillough) [inline src/RevCPU.cc:830]: I will have to look at home

File: src/RevProc.cc [hunk 4]
```diff
@@ -1905,36 +1885,55 @@ bool RevProc::ClockTick( SST::Cycle_t currentCycle ){
...
       Pipeline.front().second.cost++;
     }
   }
-  if( GetPC() == 0x00ull ){
+  // Check for completion states and new tasks
+  // @DAVE: I feel like the below two blocks of code have some redundancy but its working and
+  //        if I comment out the first block (HartToDecode based checks) the AMO tests fail
+  //        so there must be something I'm not understanding about the two. If there isn't redundancy
+  //        great, I'll uncomment this little letter I left you
+  if( RegFile->GetPC() == 0x00ull ){
```
Comment (leekillough) [inline src/RevProc.cc:1899]: `0` can always be compared against another type and will be automatically converted. It is never lossy. So something `0x000000000llu` (extreme example) is unnecessary and can be simply `0`.

**Review by leekillough — COMMENTED** [review]
I have a few other spots to review

File: include/RevProc.h [hunk 5]
```diff
@@ -190,8 +184,23 @@ class RevProc{
...
+  ///< RevProc:
+  void UpdateStatusOfHarts();
+
+  ///< RevProc: Returns the id of an idle hart (or _INVALID_HART_ID_ if none are idle)
+  unsigned FindIdleHartID() const ;
+
+  ///< RevProc: Returns true if all harts are available (ie. There is nothing executing on this Proc)
+  /// @DAVE: Can you confirm that using this function to determine if a Proc can be disabled (called in RevCPU)
+  ///        will work with the CoProc?
+  size_t HasNoBusyHarts() const { return IdleHarts == ValidHarts; }
```
[the remark this answers is not in the document]
Comment (donofrio) [reply include/RevProc.h:200]: I'll answer your question in a few ways... hopefully this helps.  

A thread should go to DONE state (as I understand it) once all instructions have been retired (similar to the `return` from `main`) and the `RevCoProc::IsDone() `returns true. 

A thread could certainly be Idle if it is waiting on a memory request or other dependency, this would be independent of the CoProc state.  It is also possible that the CoProc has forced a Hart to stall on Proc

A Proc could possibly be safely be disabled regardless of CoProc state **if** the CoProc is being clocked independently.  That is still a risky proposition as the CoProc can see and manipulate state within Proc and that is really not possible if the Proc is not being clocked.  The best strategy is to, as the general case, assume that the Proc cannot be disabled unless CoProc::IsDone() returns true.

File: include/RevProc.h [hunk 6]
```diff
@@ -213,17 +228,11 @@ class RevProc{
...
-
-  // Checks to see if a given HartID has an assigned thread by checking if the threadid on that Hart still exists in the AssignedThreads map
-  bool HartHasThread(unsigned HartID) const { return ( AssignedThreads.find(Harts.at(HartID)->GetAssignedThreadID()) != AssignedThreads.end() );}
-
   // Function pointer to the GetNewThreadID function in RevCPU (monotonically increasing thread ID counter)
   std::function<uint32_t()> GetNewThreadID;
 
   // If a given assigned thread experiences a change of state, it sets the corresponding bit
-  std::queue<std::shared_ptr<RevThread>> ThreadsThatChangedState; ///< RevProc: used to signal to RevCPU that the thread assigned to HART has changed state
+  std::queue<std::unique_ptr<RevThread>> ThreadsThatChangedState; ///< RevProc: used to signal to RevCPU that the thread assigned to HART has changed state
```
Comment (donofrio) [inline include/RevProc.h:235]: Kept the same name, eh? ;)

File: include/RevThread.h [hunk 7]
```diff
@@ -76,58 +46,100 @@ class RevThread {
...
+  // Arguments could be made to keep it here or put it there
+  void SetLSQueue(std::shared_ptr<std::unordered_map<uint64_t, MemReq>> lsq){
+    VirtRegState->SetLSQueue(lsq);
+  }
+
+  // @DAVE: Potentially move these to RevHart and have them set it when it's assigned?
+  // Arguments could be made to keep it here or put it there
+  void SetMarkLoadComplete(std::function<void(const MemReq&)> func){
+    VirtRegState->SetMarkLoadComplete(func);
+  }
```
Comment (donofrio) [inline include/RevThread.h:86]: A LSQ manages requests for multiple threads/harts but is tied to a single Proc, and as a Hart is also tied to a Proc it may make more sense to put it in RevHart, especially if threads can migrate between Procs - having LSQ in RevHart would remove the chance of a thread ever referencing a LSQ outside the Proc it is are running on.

File: src/RevProc.cc [hunk 8]
```diff
@@ -1905,36 +1885,55 @@ bool RevProc::ClockTick( SST::Cycle_t currentCycle ){
...
       // could not retire the instruction, bump the cost
       Pipeline.front().second.cost++;
     }
   }
-  if( GetPC() == 0x00ull ){
+  // Check for completion states and new tasks
+  // @DAVE: I feel like the below two blocks of code have some redundancy but its working and
+  //        if I comment out the first block (HartToDecode based checks) the AMO tests fail
+  //        so there must be something I'm not understanding about the two. If there isn't redundancy
+  //        great, I'll uncomment this little letter I left you
```
Comment (donofrio) [inline src/RevProc.cc:1898]: Dear Ryan,

Thank you for your letter. There is some redundancy - in order to be done we need to check that there is nothing stalled waiting to decode AND that even if there is nothing to decode there is no thread with an instruction currently in flight causing us to exit early. I am curious, what if you add ` HartHasNoDependencies(HartToDecodeID)` as an additional condition in the 2nd `if` does that work?

` if( HartToExecID != _REV_INVALID_HART_ID_ && !IdleHarts[HartToExecID]
        && HartHasNoDependencies(HartToExecID)
        &&  HartHasNoDependencies(HartToDecodeID)
        && ( (nullptr == coProc) || (coProc && coProc->IsDone() )
        ) ){
      std::unique_ptr<RevThread> ActiveThread = PopThreadFromHart(HartToDecodeID);
      ActiveThread->SetState(ThreadState::DONE);
      HartsClearToExecute[HartToExecID] = false;
      HartsClearToDecode[HartToExecID] = false;
      IdleHarts[HartToExecID] = true;
      ThreadsThatChangedState.emplace(std::move(ActiveThread));
    }`

File: src/RevCPU.cc [hunk 9]
```diff
@@ -770,128 +729,122 @@ void RevCPU::CheckBlockedThreads(){
...
     }
-    // Pop the thread that changed state
-    Procs[ProcID]->GetThreadsThatChangedState().pop();
+    // If the Proc does not have any threads assigned to it, there is no work to do, it is disabled
+    if( Procs[ProcID]->HasNoBusyHarts() ){
+      Enabled[ProcID] = false;
+    }
+    // Remove the pointer to the thread as it was handled
+    ThreadsThatChangedState.pop();
   }
```
Comment (donofrio) [inline src/RevCPU.cc:825]: At some point we may need to create Current vs. Next state for threads, but I think given the separate ClockTick functions in RevProc and RevCPU, we should be OK as long as we are careful in how the updates are done

File: include/RevProc.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (rkabrick) [reply include/RevProc.h:195]: > Why `size_t` return type instead of `bool`?

Isn't that a great question 😂

[the remark this answers is not in the document]
Comment (rkabrick) [reply include/RevProc.h:195]: > I'll answer your question in a few ways... hopefully this helps.
> 
> A thread should go to DONE state (as I understand it) once all instructions have been retired (similar to the `return` from `main`) and the `RevCoProc::IsDone() `returns true.
> 
> A thread could certainly be Idle if it is waiting on a memory request or other dependency, this would be independent of the CoProc state. It is also possible that the CoProc has forced a Hart to stall on Proc
> 
> A Proc could possibly be safely be disabled regardless of CoProc state **if** the CoProc is being clocked independently. That is still a risky proposition as the CoProc can see and manipulate state within Proc and that is really not possible if the Proc is not being clocked. The best strategy is to, as the general case, assume that the Proc cannot be disabled unless CoProc::IsDone() returns true.

So a thread won't be removed from a Hart until all Scoreboard entries are false. When the CoProc flags a scoreboard, is it independent of the Hart's scoreboard? Also is there state beyond just the scoreboard that the CoProc can manipulate/see into?

File: src/RevCPU.cc [hunk 10]
```diff
@@ -676,57 +660,39 @@ bool RevCPU::clockTick( SST::Cycle_t currentCycle ){
...
-  uint32_t TID = ThreadToInit->GetThreadID();
-  // Check if this ThreadID has already been assigned... if so... something has gone horribly wrong
-  // print out all threads
-  auto it = Threads.find(TID);
-  if( it != Threads.end() && it->second->GetState() != ThreadState::START ){
-    output.fatal(CALL_INFO, 99, "Error: ThreadID %" PRIu32 " has already been assigned... this is a bug.\n", TID);
-  }
-  output.verbose(CALL_INFO, 4, 0, "Initializing Thread %" PRIu32 "\n", TID);
+  RegState->SetX(RevReg::gp, gp);
+  RegState->SetX(RevReg::s0, gp); // This can be used as the fp reg per ABI
```
[the remark this answers is not in the document]
Comment (rkabrick) [reply src/RevCPU.cc:669]: In your `enum class` is it possible to have an `fp` register point to the same value as the `s0` register? Or is it a single key only

File: src/RevCPU.cc [hunk 11]
```diff
@@ -676,57 +660,39 @@ bool RevCPU::clockTick( SST::Cycle_t currentCycle ){
 
 // Initializes a RevThread object.
 // - Moves it to the 'Threads' map
-// - Adds it's ThreadID to the ThreadQueue to be scheduled
-void RevCPU::InitThread(std::shared_ptr<RevThread>& ThreadToInit){
-
+// - Adds it's ThreadID to the ReadyThreads to be scheduled
+void RevCPU::InitThread(std::unique_ptr<RevThread> ThreadToInit){
+  // @Lee, can this be a reference? I'm not looking to transfer ownership of the regfile pointer
```
[the remark this answers is not in the document]
Comment (rkabrick) [reply src/RevCPU.cc:665]: Alright I changed it to &&. Interesting discussion though. Thank you for the insight!

File: src/RevCPU.cc [hunk 12]
```diff
@@ -676,57 +660,39 @@ bool RevCPU::clockTick( SST::Cycle_t currentCycle ){
...
-
-  // Put the thread in the Proc's assigned threads list
-  AssignedThreads.at(ProcID).emplace(ThreadID, Thread);
-
-  Procs[ProcID]->AssignThread(ThreadID);
+// Assigns a RevThred to a specific Proc which then loads it into a RevHart
+// This should not be called without first checking if the Proc has an IdleHart
+void RevCPU::AssignThread(std::unique_ptr<RevThread> ThreadToAssign, unsigned ProcID){
+  output.verbose(CALL_INFO, 4, 0, "Assigning Thread %" PRIu32 " to Processor %" PRIu32 "\n", ThreadToAssign->GetID(), ProcID);
+  Procs[ProcID]->AssignThread(std::move(ThreadToAssign));
```
[the remark this answers is not in the document]
Comment (rkabrick) [reply src/RevCPU.cc:680]: In this case, you are transferring ownership of the `ThreadToAssign` to a `RevProc`. So it will not be used again until the `std::unique_ptr<RevThread>` is returned via the `ThreadsThatChangedState` handling. Does this affect your comment?

File: include/RevProc.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (donofrio) [reply include/RevProc.h:195]: At the moment the scoreboard and the register file are the only things the CoProc can manipulate.  It is possible, however, for a Hart to have a clear scoreboard and the CoProc to have outstanding work

File: src/RevCPU.cc (hunk 10 above)
[the remark this answers is not in the document]
Comment (leekillough) [reply src/RevCPU.cc:669]: > In your `enum class` is it possible to have an `fp` register point to the same value as the `s0` register? Or is it a single key only

Already `fp` registers and `x` (gpr) registers have tags which are the same values. What `fp` or `gpr` register do you want added/changed?

The names of the registers are their ABI names explicitly specified in the RISC-V Architecture SPEC (like `a0`, `sp`, `gp`, `ra`). This is intentional because those names are more familiar. Is there a name you need which is not a part of the ABI naming convention?

You can use integers instead of `RevReg` enums to give a raw register number, which is how `rs1`, `rs2`, `rs3` and `rd` are handled. The functions taking registers are templated to accept either integers or `RevReg` and are explicitly cast to `size_t`  before using them as a raw register number.

File: src/RevCPU.cc (hunk 11 above)
[the remark this answers is not in the document]
Comment (leekillough) [reply src/RevCPU.cc:665]: In the old code you were constructing a new `unique_ptr` for the function argument, and because `unique_ptr` has no copy constructor, only a move constructor, you had to call the function with `std::move(arg)`. Once in the function, the argument/parameter local variable has ownership of the `unique_ptr`. When you transfer it to its new destination, you have to use `std::move` again.

With `&&` rvalue reference for the function parameter, you are requiring that it only bind to rvalues, so lvalues have to be converted to rvalues by using `std::move` at the call site. Within the function, the parameter name acts as an lvalue to the referent. You have to use `std::move` again to transfer it to its new destination.

The difference between old and new is that a local temporary `unique_ptr` and moves in and out of it were avoided -- now you just use a rvalue reference for the parameter. It still only accepts rvalue arguments, and it still must be converted to a rvalue when it is transferred to somewhere else.

File: src/RevCPU.cc (hunk 10 above)
[the remark this answers is not in the document]
Comment (leekillough) [reply src/RevCPU.cc:669]: > In your `enum class` is it possible to have an `fp` register point to the same value as the `s0` register? Or is it a single key only

Sorry I misunderstood. `fp` is the official ABI name of frame pointer, which is why it's used in the `RevReg` enum.

I was referring to your **comment** after `//`. "Frame pointer" should be used in the comment instead of `fp` which is ambiguous and means "floating point" much more often.

[the remark this answers is not in the document]
Comment (leekillough) [reply src/RevCPU.cc:669]: "This can be used as the fp reg..." should be "This can be used as the frame pointer reg"

[the remark this answers is not in the document]
Comment (leekillough) [reply src/RevCPU.cc:669]: "fp reg" 99% of the time means "floating point register", so "frame pointer" should be worded explicitly to avoid confusion.

`RevReg::fp` is the ABI's official name for the frame pointer register. The `RevReg::` prefix makes that unambiguous. Bare "fp" used in comments almost always means floating-point.

[the remark this answers is not in the document]
Comment (leekillough) [reply src/RevCPU.cc:669]: ... and since there are "fp registers", "the fp register" is even more ambiguous.

File: include/RevProc.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (rkabrick) [reply include/RevProc.h:195]: Enough said... I'll adjust the check in RevCPU

File: src/RevCPU.cc (hunk 10 above)
[the remark this answers is not in the document]
Comment (rkabrick) [reply src/RevCPU.cc]: Ah yeah those are all very valid points

File: src/RevProc.cc (hunk 4 above)
Comment (rkabrick) [reply src/RevProc.cc:1899]: This is how it's been used throughout RevProc... Up to @jleidel . Not sure if there was a reason for doing it that way originally

File: include/RevProc.h (hunk 6 above)
Comment (rkabrick) [reply include/RevProc.h:236]: Changing to Homer

File: include/RevThread.h [hunk 13]
```diff
@@ -76,58 +46,100 @@ class RevThread {
...
 
-  // See if file descriptor exists/is owned by this thread
+  ///< See if file descriptor exists/is owned by this thread
   bool FindFD(int fd){ return fildes.count(fd); }
 
-  // Get the threads
+  // @DAVE: Potentially move these to RevHart and have them set it when it's assigned?
+  // Arguments could be made to keep it here or put it there
+  void SetLSQueue(std::shared_ptr<std::unordered_map<uint64_t, MemReq>> lsq){
+    VirtRegState->SetLSQueue(lsq);
```
[the remark this answers is not in the document]
Comment (leekillough) [reply include/RevThread.h:79]: If you a reference, you'll avoid an atomic increment/decrement of the reference count which occurs when creating and destroying the `lsq` local parameter.
```c++
  void SetLSQueue(const std::shared_ptr<std::unordered_map<uint64_t, MemReq>>& lsq){
```

File: include/RevThread.h [hunk 14]
```diff
@@ -76,58 +46,100 @@ class RevThread {
...
+  }
+
+  // @DAVE: Potentially move these to RevHart and have them set it when it's assigned?
+  // Arguments could be made to keep it here or put it there
+  void SetMarkLoadComplete(std::function<void(const MemReq&)> func){
+    VirtRegState->SetMarkLoadComplete(func);
```
[the remark this answers is not in the document]
Comment (leekillough) [reply include/RevThread.h]: Copying a `std::function` requires copying all of its captured variables, allocating memory, etc. Moving it should be faster.

File: src/RevCPU.cc (hunk 12 above)
[the remark this answers is not in the document]
Comment (leekillough) [reply src/RevCPU.cc:686]: I don't know what comment "above" I was referring to, but I think it involves the rvalue reference and move stuff, so you should change the function to:
```c++
void RevCPU::AssignThread(std::unique_ptr<RevThread>&& ThreadToAssign, unsigned ProcID){
```
The rest of the function stays the same. You are avoiding creating a temporary local `ThreadToAssign` and are using a rvalue reference to the replacement directly.

File: include/RevThread.h (hunk 7 above)
Comment (rkabrick) [reply include/RevThread.h]: Moved... the LSQueue & MarkLoadComplete are now parts of the RevHart constructor. Everything passes but please review the new implementation in RevHart

File: src/RevProc.cc (hunk 8 above)
Comment (rkabrick) [reply src/RevProc.cc:1899]: You're a legend! Running tests now and it's looking like that works

Comment (rkabrick) [reply src/RevProc.cc:1899]: nvm those tests time out

File: src/RevCPU.cc (hunk 9 above)
Comment (rkabrick) [reply src/RevCPU.cc:829]: That seems like something that is definitely possible to achieve with the current implementation (ie. we haven't coded ourselves into a corner) so if it's cool with you I'll leave it as is in this PR and down the road we can revisit

File: src/RevCPU.cc (hunk 3 above)
Comment (rkabrick) [reply src/RevCPU.cc:834]: Let me know

File: src/RevCPU.cc [hunk 15]
```diff
@@ -770,128 +729,122 @@ void RevCPU::CheckBlockedThreads(){
...
+    }
+    // Remove the pointer to the thread as it was handled
+    ThreadsThatChangedState.pop();
   }
   return;
 }
 
+void RevCPU::InitMainThread(uint32_t MainThreadID, const uint64_t StartAddr){
+  // @Lee: Is there a better way to get the feature info?
+  std::unique_ptr<RevRegFile> MainThreadRegState = std::make_unique<RevRegFile>(Procs[0]->GetRevFeature());
```
Comment (leekillough) [inline src/RevCPU.cc:835]: `Procs[0]` seems to be arbitrary. Is there a way to figure out which `Procs` to use here?

File: src/RevProc.cc (hunk 4 above)
Comment (leekillough) [reply src/RevProc.cc:1900]: Being the minimalist I am, I rarely compare equal or non-equal to 0 -- I would use:
```c++
if( !RegFile->GetPC() ){
```

File: test/threading/pthreads/pthread-arg-passing/pthread-arg-passing.c [hunk 16]
```diff
@@ -1,13 +1,19 @@
 #include "../../../../common/syscalls/syscalls.h"
-#define assert(x) if (!(x)) { asm(".byte 0x00"); asm(".byte 0x00"); asm(".byte 0x00"); asm(".byte 0x00"); }
+#define assert(x)                                                              \
+  if (!(x)) {                                                                  \
+    asm(".byte 0x00");                                                         \
```
Comment (leekillough) [inline test/threading/pthreads/pthread-arg-passing/pthread-arg-passing.c:4]: `asm(".word 0");` instead of four `asm('.byte 0x00");` would be simpler

File: src/RevCPU.cc (hunk 15 above)
Comment (rkabrick) [reply src/RevCPU.cc:835]: It is semi-arbitrary. I used 0 because we ALWAYS know that there will be at least one Proc (Core ;)). Further, we also know that the main thread will always end up on the first core. 

I still don't like it... but currently the RevFeature is created inside of RevProc so that's the only way I can think to gather the info.

**Review by leekillough — CHANGES_REQUESTED** [review]
See https://github.com/tactcomplabs/rev/pull/185

File: src/RevCPU.cc (hunk 15 above)
Comment (donofrio) [reply src/RevCPU.cc:830]: Right now we assume a homogeneous CPU, so this is OK.  To get to a heterogeneous CPU (like a big.LITTLE) we'll need to do some rearchitecting and the CPU will need to have some description of what features (RV and other) are in which Procs. I don't think we need to solve that here, or hold up your PR for that reason
